Src/Base/AMReX_Array4.H

```diff
@@ -56,212 +57,373 @@ namespace amrex {
+            : p((T*)(rhs.p + start_comp*rhs.nstride[dim-2])),
               nstride(rhs.nstride),
               begin(rhs.begin),
-              end(rhs.end),
-              ncomp(rhs.ncomp-start_comp)
-        {}
+              end(rhs.end)
+        {
+            begin[3] = 0;
+            end[3] = rhs.end[3] - start_comp;
+        }
 
         template <class U,
-                  std::enable_if_t
+            std::enable_if_t
                   <std::is_same_v<std::remove_const_t<T>,
-                                std::remove_const_t<U>>,int> = 0>
+                                std::remove_const_t<U>> && (dim == 4),int> = 0>
         AMREX_GPU_HOST_DEVICE
-        constexpr Array4 (Array4<U> const& rhs, int start_comp, int num_comps) noexcept
-            : p((T*)(rhs.p+start_comp*rhs.nstride)),
-              jstride(rhs.jstride),
-              kstride(rhs.kstride),
+        constexpr ArrayND (ArrayND<U, dim> const& rhs, int start_comp, int num_comp) noexcept
+            : p((T*)(rhs.p + start_comp*rhs.nstride[dim-2])),
               nstride(rhs.nstride),
               begin(rhs.begin),
-              end(rhs.end),
-              ncomp(num_comps)
-        {}
+              end(rhs.end)
+        {
+            begin[3] = 0;
+            end[3] = num_comp;
+        }
 
         AMREX_GPU_HOST_DEVICE
         explicit operator bool() const noexcept { return p != nullptr; }
 
-        template <class U=T, std::enable_if_t<!std::is_void_v<U>,int> = 0>
-        [[nodiscard]] AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
-        U& operator() (int i, int j, int k) const noexcept {
-#if defined(AMREX_DEBUG) || defined(AMREX_BOUND_CHECK)
-            index_assert(i,j,k,0);
-#endif
-            return p[(i-begin.x)+(j-begin.y)*jstride+(k-begin.z)*kstride];
-        }
-
-        template <class U=T, std::enable_if_t<!std::is_void_v<U>,int> = 0>
-        [[nodiscard]] AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
-        U& operator() (int i, int j, int k, int n) const noexcept {
-#if defined(AMREX_DEBUG) || defined(AMREX_BOUND_CHECK)
-            index_assert(i,j,k,n);
-#endif
-            return p[(i-begin.x)+(j-begin.y)*jstride+(k-begin.z)*kstride+n*nstride];
-        }
-
-        template <class U=T, std::enable_if_t<!std::is_void_v<U>,int> = 0>
+        template <typename... idx, class U=T,
+            std::enable_if_t<!std::is_void_v<U> && detail::ArrayNDIndexCheck_impl_v<dim, idx...>,int> = 0>
         [[nodiscard]] AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
-        T* ptr (int i, int j, int k) const noexcept {
+        U& operator() (idx... i) const noexcept {
 #if defined(AMREX_DEBUG) || defined(AMREX_BOUND_CHECK)
-            index_assert(i,j,k,0);
+            if constexpr (sizeof...(i) == dim) {
+                index_assert(i...);
+            } else if constexpr (sizeof...(i) == dim - 1) {
+                // When number of indices is one less than dimension, we assume the last index is zero
```

Comment:
In this case the last index should not be zero but effectively `begin[dim-1]`. If it were zero, the index calculation would still add instructions to subtract `begin[dim-1]`, multiply by the stride, and add the result to the offset. Currently with Array4 this is not done. To guarantee the absence of the extra instructions, there should be two versions of `get_offset` where one of them leaves out the calculation for the last index.